### ISSUE TITLE
Data messages

### ISSUE THREAD

**Issue by dbrgn** [saltyrtc/saltyrtc-meta#34]
- [ ] Drop offer / answer / candidates message types
- [ ] Add generic data message type as the only p2p message type after the handshake
- [ ] Specify webrtc / ortc message types as extension

In TypeScript:

``` typescript
interface Message {
  type: string,
}

interface Data extends Message {
  type: 'data',
  data_type: string,
  data: any,
}
```

The `data_type` field allows us to register event handlers for specific data messages only.

Open question: Should we make the `data_type` field optional? If it's not specified, then the use of `onData('type', function(msg) { ... })` type handlers is not possible in a generic way.

Related to #33.

**Comment by lgrahl** [comment]
I think the question of having `data_type` required or optional depends on how events will be registered. In the W3C-way (`instance.onsomeevent = function(...) { ... }`) or in another way? Personally, I feel we should be as close to W3C specs as possible. But I'd also like to have the alternative form of `instance.onData('type', function(...) { ... })`. Any suggestions?

**Comment by dbrgn** [comment]
Generally I'd go with an `onData(event, handler)` approach. It's more flexible. You can use it as usual by registering a single handler or you can register multiple handlers.

On the other hand, if we have `onConnected` and `onConnectionError` with a traditional model, that would be confusing.

Regarding the W3C, they also document (not sure if that can be considered endorsing) "event-emitter style" handlers in their wiki, e.g. here: https://www.w3.org/wiki/Handling_events_with_JavaScript (`addEvent`). I think we should strive for the best usability, not for the most W3C-esque API.

(Another example: WebRTC started using promise based APIs, even though most existing APIs use callbacks.)

```javascript

**Comment by dbrgn** [comment]
Generally I'd go with an `onData(event, handler)` approach. It's more flexible. You can use it as usual by registering a single handler or you can register multiple handlers.

On the other hand, there's currently `onConnected` and `onConnectionError` with a traditional model, which makes the situation confusing.

Regarding the W3C, they also document (not sure if that can be considered endorsing) "event-emitter style" handlers in their wiki, e.g. here: https://www.w3.org/wiki/Handling_events_with_JavaScript (`addEvent`). I think we should strive for the best usability, not for the most W3C-esque API.

(Another example: WebRTC started using promise based APIs, even though most existing APIs use callbacks.)

So we should probably choose either this API:

``` typescript
onConnected() = function() { ... };
onDataMessage(msg) = function(msg) { ... };
onConnectionError(ev) = function(ev) { ... };
onConnectionClosed(ev) = function(ev) { ... };
```

...or this one:

``` typescript
type EventType = 'connected' | 'connection-closed' | 'connection-error';

onEvent(eventType: EventType, handler);
onData(dataType: string, handler);
```

The first one is simpler, the second one is more flexible and allows the user to register callbacks for specific data messages (e.g. an offer or an answer). If we choose the first, every user will have to implement either a state machine or his/her own event emitter.

**Comment by dbrgn** [comment]
On second thought, with the second approach we'd also need ways to de-register events and to run events only once. If there's only one event type, that's easy: `on(...)`, `off(...)`, `once(...)`. But with two event types (event or data) that API would look stupid (`offEvent`, `onceEvent` etc).

**Comment by dbrgn** [comment]
_(Comment moved to https://github.com/saltyrtc/saltyrtc-client-js/issues/16)_

**Comment by lgrahl** [comment]
I guess we should have discussed this in the `saltyrtc-client-js` repo. For other readers: The discussed API will not be part of the SaltyRTC protocol specification. @dbrgn could you create a new issue for the proposed changes to the signalling messages and close this issue?

**Comment by dbrgn** [comment]
Yes, I just realized that as well a few minutes ago.

Regarding this issue, I'd suggest to specify the `data` message type as follows:

```
{ type: 'data', data: any, [data_type: string] }
```

**Comment by lgrahl** [comment]
@dbrgn Should we stick with that or just have new message types? I'd rather have new message types because we'd avoid nesting that way.

**Comment by dbrgn** [comment]
Data messages make the processing simpler, because the decryption code can simply drop anything with an unknown message type, and pass all data messages to the tasks.

On the other hand you're right about unnecessary nesting from a message size perspective.

I still think I'd stick with the data messages though, they give you more flexibility and simplify implementation. But if you decide to go with new message types I'd be OK with that too.

**Comment by lgrahl** [comment]
How about a compromise: The message will still be flat, but...

```
{
  "type": "data",
  "task_type": "whatever",
  ...
}
```

Would that help?

**Comment by dbrgn** [comment]
That will probably still cause issues with auto-deserialization into objects. The version where everything is nested under the `data` key can be deserialized with a set of known top level keys, but in the version you proposed, the present keys aren't known yet at decode time.

In other words, the Java client can deserialize the nested version as `{type: String, data_type: String, data: Object}` but the flat version cannot be deserialized into an object because the attributes aren't known in advance.

**Comment by dbrgn** [comment]
Note that we could implement task message types as flat messages without "type=data". It just makes the implementation a bit more complicated, because these message types need to be registered depending on the supported and/or negotiated tasks.

Therefore I'd either go with nested data messages, or instead without any data messages (specifying all task messages on "top level"). In the latter case we'd lose the possibility of implementing simple protocols without any tasks at all.

**Comment by lgrahl** [comment]
In that case, I'd rather go without any 'data' messages at all and see how messy it gets (hopefully not that badly). There'll always be a task after C2C authentication is complete. Did you mean something else?

**Comment by dbrgn** [comment]
With the data message, someone could transmit data without any formal task. It's not meant for that, but it does not require any change in the client libraries.

If our own tasks (webrtc, ortc, etc) are implemented directly in the client libraries, a user that wants to create a custom task must change the client libraries.

I'd have to see if there's a clean way to make the tasks separate projects, with an API in the regular saltyrtc to register these tasks. Would probably be cleaner.

**Comment by lgrahl** [comment]
> I'd have to see if there's a clean way to make the tasks separate projects, with an API in the regular saltyrtc to register these tasks. Would probably be cleaner.

I think that would be the best approach. Separating the WebRTC and the ORTC task from the SaltyRTC client. :+1: This is the way it's going to be for the spec, too.

And I'm absolutely fine if there's a task that does nothing but relay data. Early data transmission for WebRTC and ORTC would be great, too.

**Comment by lgrahl** [comment]
I'll close this one. Early data transmission for ORTC and WebRTC tasks can be discussed in #55. I'll be fine with a separate _Data Only_ task once someone has a use case for it - just open a new issue. :)
